url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/ACS.java

```diff
@@ -0,0 +1,1521 @@
+	 * performed by default.*/
+	public static boolean GLOBAL_PERFORM_EXTRACTION = true;
+	/**Specifies whether or not extraction should be used. Extraction is 
+	 * performed by default.*/
+	public boolean PERFORM_EXTRACTION = GLOBAL_PERFORM_EXTRACTION;
+	/**Specifies whether or not bottom level learning should be performed. Learning is 
+	 * performed by default.*/
+	public static boolean GLOBAL_PERFORM_BL_LEARNING = true;
+	/**Specifies whether or not bottom level learning should be performed. Learning is 
+	 * performed by default.*/
+	public boolean PERFORM_BL_LEARNING = GLOBAL_PERFORM_BL_LEARNING;
+	/**Specifies whether or not refinement should be used for RER rules. Refinement is 
+	 * performed by default.*/
+	public static boolean GLOBAL_PERFORM_REFINEMENT_RER = true;
+	/**Specifies whether or not refinement should be used for RER rules. Refinement is 
+	 * performed by default.*/
+	public boolean PERFORM_REFINEMENT_RER = GLOBAL_PERFORM_REFINEMENT_RER;
+	/**Specifies whether or not refinement should be used for IRL rules. Refinement is 
+	 * performed by default.*/
+	public static boolean GLOBAL_PERFORM_REFINEMENT_IRL = true;
+	/**Specifies whether or not refinement should be used for IRL rules. Refinement is 
+	 * performed by default.*/
+	public boolean PERFORM_REFINEMENT_IRL = GLOBAL_PERFORM_REFINEMENT_IRL;
+	
+	/**The match discount factor.*/
+	public static double GLOBAL_MATCH_DISCOUNT = .9;
+	/**The match discount factor.*/
+	public static double MATCH_DISCOUNT = GLOBAL_MATCH_DISCOUNT;
+	
+	/**The subset of possible actions that were involved in the last action decision making stochastic selection.*/
+	private Collection <AbstractAction> shortlist;
+	
+	/**
+	 * Minimally initializes the ACS. This constructor builds a "frame" for this instance of
+	 * the ACS from which all desired components can be attached. During initialization this
+	 * instance of the ACS will attach itself to the CLARION agent you specify.
+	 * <p>
+	 * Note that simply calling this constructor is not sufficient enough to setup the ACS.
+	 * At the very least you MUST also add at least one neural net to the bottom level of
+	 * the ACS along with all of the actions associated with the output layer of that net.
+	 * @param Agent The agent to which the ACS is being attached.
+	 */
+	public ACS (CLARION Agent) 
+	{
+		super(Agent);
+		addAction(new ExternalAction("DO_NOTHING_EXTERNAL"));
+		addAction(new WorkingMemoryAction("DO_NOTHING_WM"));
+		addAction(new GoalAction("DO_NOTHING_GOAL"));
+	}
+	
+	/**
+	 * Chooses an action given the current input (which includes sensory information, the 
+	 * current goal, and any chunks in working memory) and the previously chosen action.
+	 * @param CurrInput The current input to the ACS.
+	 * @param TimeStamp The current time stamp.
+	 * @return The chosen action.
+	 */
+	protected AbstractAction chooseAction (DimensionValueCollection CurrInput, long TimeStamp)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 298-437 into a separate method.